File: src/main/java/com/amrit/futsal/service/ReportService.java
package com.amrit.futsal.service;

import com.amrit.futsal.entity.Booking;
import com.amrit.futsal.entity.Report;
import com.amrit.futsal.entity.User;
import com.amrit.futsal.repository.ReportRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
public class ReportService {

    private final ReportRepository reportRepository;
    private final BookingService bookingService;

    @Autowired
    public ReportService(ReportRepository reportRepository, BookingService bookingService) {
        this.reportRepository = reportRepository;
        this.bookingService = bookingService;
    }

    public Report createReport(Report report) {
        return reportRepository.save(report);
    }

    public Optional<Report> getReportById(UUID reportId) {
        return reportRepository.findById(reportId);
    }

    public List<Report> getReportsByOwnerId(UUID ownerId) {
        return reportRepository.findByOwnerId(ownerId);
    }
    
    public List<Report> getReportsByType(Report.ReportType reportType) {
        return reportRepository.findByReportType(reportType);
    }
    
    public List<Report> getReportsByDateRange(LocalDateTime start, LocalDateTime end) {
        return reportRepository.findByGeneratedAtBetween(start, end);
    }
    
    public List<Report> getReportsByOwnerAndType(UUID ownerId, Report.ReportType reportType) {
        return reportRepository.findByOwnerIdAndReportType(ownerId, reportType);
    }
    
    /**
     * Generates a revenue report for a specific owner
     */
    public Report generateRevenueReport(User owner) {
        // Implementation will depend on how you want to generate the report
        Report report = new Report();
        report.setOwner(owner);
        report.setReportType(Report.ReportType.REVENUE);
        
        // Save the report
        return reportRepository.save(report);
    }
    
    /**
     * Generates a bookings report for a specific owner
     */
    public Report generateBookingsReport(User owner) {
        // Implementation will depend on how you want to generate the report
        Report report = new Report();
        report.setOwner(owner);
        report.setReportType(Report.ReportType.BOOKINGS);
        
        // Save the report
        return reportRepository.save(report);
    }
    
    /**
     * Generates a customers report for a specific owner
     */
    public Report generateCustomersReport(User owner) {
        // Implementation will depend on how you want to generate the report
        Report report = new Report();
        report.setOwner(owner);
        report.setReportType(Report.ReportType.CUSTOMERS);
        
        // Save the report
        return reportRepository.save(report);
    }
    
    public void deleteReport(UUID reportId) {
        reportRepository.deleteById(reportId);
    }
}

File: frontend/src/services/reportService.ts
import apiClient from './api';
import { Report, ReportType } from '../types';

export const reportService = {
  async getReportById(id: string): Promise<Report> {
    const response = await apiClient.get<Report>(`/reports/${id}`);
    return response.data;
  },

  async getMyReports(): Promise<Report[]> {
    const response = await apiClient.get<Report[]>('/reports/me');
    return response.data;
  },

  async getReportsByOwner(ownerId: string): Promise<Report[]> {
    const response = await apiClient.get<Report[]>(`/reports/owner/${ownerId}`);
    return response.data;
  },

  async getReportsByType(reportType: ReportType): Promise<Report[]> {
    const response = await apiClient.get<Report[]>(`/reports/type/${reportType}`);
    return response.data;
  },

  async getReportsByDateRange(startDate: string, endDate: string): Promise<Report[]> {
    const response = await apiClient.get<Report[]>('/reports/date-range', {
      params: { start: startDate, end: endDate },
    });
    return response.data;
  },

  async generateRevenueReport(
    startDate?: string,
    endDate?: string,
    ownerId?: string
  ): Promise<Report> {
    const endpoint = ownerId ? `/reports/generate/revenue/${ownerId}` : '/reports/generate/revenue';
    const response = await apiClient.post<Report>(endpoint, null, {
      params: { startDate, endDate },
    });
    return response.data;
  },

  async generateBookingsReport(
    startDate?: string,
    endDate?: string,
    ownerId?: string
  ): Promise<Report> {
    const endpoint = ownerId ? `/reports/generate/bookings/${ownerId}` : '/reports/generate/bookings';
    const response = await apiClient.post<Report>(endpoint, null, {
      params: { startDate, endDate },
    });
    return response.data;
  },

  async generateCustomersReport(
    startDate?: string,
    endDate?: string,
    ownerId?: string
  ): Promise<Report> {
    const endpoint = ownerId ? `/reports/generate/customers/${ownerId}` : '/reports/generate/customers';
    const response = await apiClient.post<Report>(endpoint, null, {
      params: { startDate, endDate },
    });
    return response.data;
  },

  async deleteReport(id: string): Promise<void> {
    await apiClient.delete(`/reports/${id}`);
  },
};
